### DISCUSSION TITLE
Stepper Motor in accessory

### DISCUSSION THREAD

**Discussion by Diederik98** [HomeSpan/HomeSpan#58] [Q&A]
Hi, I was trying out this awesome library but got stuck when using a stepper motor (AccelStepper library).
When I run the code for my stepper in its own sketch, this executes without stuttering. 
However when I try this within your library, this stuttering will take place for several ms. 
Another thing I tried is to attach my stepper to core 0, but then I'm still experiencing the motor problems.
It's however still not as smooth as the example running in its own sketch. 

Do you think it is possible to get a smooth behavior? 

Thanks in advance!

**Comment by HomeSpan** [comment]
I'm unfamiliar with this library so I pulled the zip file and reviewed the code, but please take these comments as high-level observations.  Someone more familiar with the details of the code may have a different view.

This library seems to require the processor to continuously poll various routines and uses time stamps and delays to determine when to change stepper motor pins high or low.   This approach is likely to cause exactly the symptoms you see when used in a complex program.

Ideally, time-sensitive or repetitive tasks should be implemented at the hardware level whenever possible, not in software, for these very reasons.  In fact, I developed HomeSpan because I needed to control RF appliances and wanted to generate such signals in hardware using some form of PWM.  This is possible in both the Arduino and the ESP32.   I was surprised to find the ESP32 makes this even easier since it has a built in signal generator peripheral that outputs pre-programmed pulses.  Note that even the blinking status light used in HomeSpan is controlled via hardware timers as opposed to repeated calls by the software.

My guess is that the speed at which the main loop runs in HomeSpan is just too long to reliably poll stepper motor logic without modifying the stepper motor software to instead utilize some of the ESP32's pulse-generation hardware, timers, and interrupts.  Running in Core 0 likely will yield the same problems since that core is very busy maintaining the WiFi and other central services.

One possibility might be to run the stepper motor logic as a separate task.  I've not done this before so not sure what it would entail (and not sure it would solve the problem), but in looking at some other stepper motor libraries on the web, it seems that this approach has been used (though this is not the same as changing the core logic of the library to use hardware timers and interrupts).

Sorry not to be able to provide an easy solution for this issue.  I'll put this on the list of future things to explore, since driving a stepper motor with an ESP32 controlled by HomeKit is a very logical use of the platform.

**Comment by HomeSpan** [comment]
Followup - I found a library you may want to consider:  https://github.com/gin66/FastAccelStepper.  According to the README, this library uses the MCPWM hardware peripheral built into the ESP32 to control the stepper motor.  This may (should) provide a much better experience.

**Reply by Diederik98** [reply]
Thanks a lot for your detailed explanation. I'll give FastAccelStepper a try.

**Comment by blaisebarrette** [comment]
When using such library that needs to instantiate an object, where or how do you instantiate it? Before the struct, inside the struct before the constructor, inside the constructor?

I tried a few things with no success. The stepper object is never accessible from the HomeSpan loop.

Thanks!

**Comment by HomeSpan** [comment]
There are numerous ways of doing this in C++.  Assume I have a structure or class, **MySwitch**, that takes three arguments in the constructor.  I usually do one of the following depending on the circumstances:

1. Create a global instance that is defined outside of the Arduino `setup()` function so that it can be used anywhere in the sketch.  However, rather the *instantiate* it outside of `setup()`, which consumes limited stack memory, I define a global (i.e. outside of `setup()` *pointer* to the class, and then instantiate *within* `setup()` like this:

```C++

MySwitch *switch1;   // define a pointer to the structure or class with global scope (i.e. it can be used anywhere in the sketch)

void setup(){
  switch1 = new MySwitch(x,y,z);   // instantiate the object using its constructor
}

void loop(){
  switch1->doSomething();   // call doSomething method for switch1 object
}
```

2. Create the object *inside* a HomeSpan Service structure, and reference it from `loop()` by saving a pointer to the Service structure itself for later use.

```C++
struct DEV_LED : Service::LightBulb {               // here is a Service structure from Example 5

  int ledPin;
  SpanCharacteristic *power;
  MySwitch switch1{x,y,z};     // note when you instantiate an object outside of a method you need to use {} instead of ()
  
  DEV_LED(int ledPin) : Service::LightBulb(){
 ....    
  }
} // end DEV_LED definition

DEV_LED *myLED;   // here we define a pointer to the DEV_LED structure outside of `setup()` so it has global scope

void setup(){

  new SpanAccessory();   
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
  
    myLED = new DEV_LED(17);   // here we instantiate DEV_LED, just as in Example 5, BUT we save the pointer for later use
}

void loop(){
  myLED->switch1->doSomething();  // by saving the pointer to DEV_LED, we can access switch1
}

**Comment by blaisebarrette** [comment]
Wow! Thank you so much for this explanation! I would have never found it  by luck like I often do with procedural code. Object oriented is something else! Clearly way sexier and I really wanna learn! Because of HomeSpan, I am so impressed, intrigued and inspired by you’re code that I subscribed to CodeBeauty’s YouTube  channel. I have already watched most of her videos. I’m starting to grasp some concepts but I’m not quite there yet. At least I’m making the effort!

That explanation is gonna be very helpful! You should consider putting this in the documentation for lesser coders like me. To bridge the gap for people that are only used to procedural programming enough to use your library without necessarily understanding all the object programming structure.

Again, thank you for taking the time to write this work of art and also taking your precious time for answering so thoroughly our questions! You kick ass!

**Comment by blaisebarrette** [comment]
I have tried FastAccelStepper with no success.
My code compiles fine using your previous explanation. 
Once uploaded, Everything is still fine. I can even pair with the Home app with no issue.
It's only when I send a call from the Home app to open or close a curtain that the ESP32 start rebooting endlessly.
I even tried two other libraries. AccelStepper and the very basic Arduino/Stepper.
All three have the same behaviour.

> ** Client #0 Connected: (22 sec) 192.168.1.80 on Socket 3/16
> In Get Accessories #0 (192.168.1.80)...
> In Get Characteristics #0 (192.168.1.80)...
> In Put Characteristics #0 (192.168.1.80)...
> Notification Request for aid=1 iid=4: true
> Notification Request for aid=1 iid=5: true
> Notification Request for aid=1 iid=6: true
> Notification Request for aid=1 iid=8: true
> In Put Characteristics #0 (192.168.1.80)...
> Recieved call to fully open the curtains
> Updating aid=1 iid=5 (okay)
> Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
> 
> Core  1 register dump:
> PC      : 0x400dd83f  PS      : 0x00060330  A0      : 0x800d3552  A1      : 0x3ffb26c0  
> A2      : 0x00000000  A3      : 0x00000000  A4      : 0x0000588d  A5      : 0x00009dc0  
> A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0xfa000000  A9      : 0x00000000  
> A10     : 0x567a0000  A11     : 0x9dc00000  A12     : 0x00000000  A13     : 0x00000008  
> A14     : 0x00000000  A15     : 0x3ffb2600  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
> EXCVADDR: 0x00000042  LBEG    : 0x40084709  LEND    : 0x40084711  LCOUNT  : 0x00000026  
> 
> 
> Backtrace:0x400dd83c:0x3ffb26c00x400d354f:0x3ffb26e0 0x400d842d:0x3ffb2700 0x400d85c2:0x3ffb27e0 0x400d2f32:0x3ffb2800 0x400dfc9d:0x3ffb2820 
> 
> 
> 
> 
> ELF file SHA256: 0000000000000000
> 
> Rebooting...

I had great hopes for FastAccelStepper library since you recommended it and it can easily be assigned to a specific core.

Here is the code for my Curtain.h file:


```
/*--- Uncoment the stepper library you wish to use ---*/
//          #define StepperSupport
//          #define AccelStepperSupport
          #define FastAccelStepperSupport
/*----------------------------------------------------*/

#ifdef StepperSupport
    #include <Stepper.h>
    Stepper *stepperMotor;
#endif

#ifdef AccelStepperSupport
    #include <AccelStepper.h>
    AccelStepper *stepperMotor;
#endif

#ifdef FastAccelStepperSupport
    #include <FastAccelStepper.h>
    #define dirPinStepper 21
    #define enablePinStepper 22
    #define stepPinStepper 23
    FastAccelStepperEngine engine = FastAccelStepperEngine();
    FastAccelStepper *stepperMotor = NULL;
#endif

struct DEV_Curtain : Service::WindowCovering { // Smart Curtains
    
    int IsMoving = 0;  // 0 = Not moving, 1 = moving down, 2 = moving up

    SpanCharacteristic* Current;        // reference to the 'Current WindowCovering State' Characteristic
    SpanCharacteristic* Target;         // reference to the 'Target WindowCovering State' Characteristic 
    SpanCharacteristic* State;          // reference to the 'position state' Characteristic
    SpanCharacteristic* Hold;           // reference to the 'Hold position' Characteristic
    SpanCharacteristic* Obstruction;    // reference to the 'Obstruction detected' Characteristic
    
    DEV_Curtain():Service::WindowCovering(){ // constructor() method
        Current=(new Characteristic::CurrentPosition(0, true))->setRange(0,100,100); // initial value of 0 means closed, 100 means fully open
        Target=(new Characteristic::TargetPosition(0, true))->setRange(0,100,100);   // initial value of 0 means closed, 100 means fully open
        State=new Characteristic::PositionState(2);                                 // initial value of 2 means stoped, 0 means closing, 1 means opening
        Hold=new Characteristic::HoldPosition(false);                               // initial value of 0, 1 means hold position, 0 is ignored (Boolean)
        Obstruction=new Characteristic::ObstructionDetected(false);                 // initial value of 0, 1 means hold position, 0 is ignored (Boolean)

    } // end constructor

    boolean update(){ // update() method
        
        if(Target->getNewVal() == 100){ // Recieved a call to fully open the Window Covering
            LOG1("Recieved call to fully open the curtains\n");
            IsMoving = 1; // Moving down (opening)
            State->setVal(1);   // 1 means opening
        }

        if(Target->getNewVal() == 0){ // Recieved a call to fully close the Window Covering
            LOG1("Recieved call to fully close the curtains\n");
            IsMoving = 2; // Moving up (closing)
            State->setVal(0);   // 0 means closing
        }

        return(true);   // return true
    } // update
    
    void setup(){
        #define limitSwitchOpen 5
        #define limitSwitchClosed 15
        pinMode(limitSwitchOpen,INPUT_PULLUP);
        pinMode(limitSwitchClosed,INPUT_PULLUP);

        #ifdef StepperSupport
            const int stepsPerRevolution = 200;  // change this to fit the number of steps per revolution
            #define motorPin1 21     // IN1 on ULN2003 ==> Blue   on 28BYJ-48
            #define motorPin2 19    // IN2 on ULN2004 ==> Pink   on 28BYJ-48
            #define motorPin3 22    // IN3 on ULN2003 ==> Yellow on 28BYJ-48
            #define motorPin4 23    // IN4 on ULN2003 ==> Orange on 28BYJ-48
            stepperMotor = new Stepper(stepsPerRevolution, motorPin1, motorPin3, motorPin2, motorPin4);
        #endif

        #ifdef AccelStepperSupport
            #define HALFSTEP  4
            #define motorPin1 21     // IN1 on ULN2003 ==> Blue   on 28BYJ-48
            #define motorPin2 19    // IN2 on ULN2004 ==> Pink   on 28BYJ-48
            #define motorPin3 22    // IN3 on ULN2003 ==> Yellow on 28BYJ-48
            #define motorPin4 23    // IN4 on ULN2003 ==> Orange on 28BYJ-48
            #define endPoint 1000.0
            stepperMotor = new AccelStepper(HALFSTEP, motorPin1, motorPin3, motorPin2, motorPin4);
            stepperMotor->setMaxSpeed(300.0);
            stepperMotor->setAcceleration(10000.0);
            stepperMotor->setSpeed(300);
            stepperMotor->disableOutputs();
            stepperMotor->moveTo(endPoint);
        #endif

        #ifdef FastAccelStepperSupport
            #define SUPPORT_CPU_AFFINITY
            engine.init(0); // Put 0 or 1 between parentesis to choose on witch core to execute task
            stepperMotor = engine.stepperConnectToPin(stepPinStepper);
            if (stepperMotor) {
                stepperMotor->setDirectionPin(dirPinStepper);
                stepperMotor->directionPinHighCountsUp();
                stepperMotor->setEnablePin(enablePinStepper);
                stepperMotor->setAutoEnable(true);
                stepperMotor->setSpeedInHz(500);       // 500 steps/s
                stepperMotor->setAcceleration(10000);    // 10 000 steps/s²
                stepperMotor->setCurrentPosition(0);
            }
        #endif

        if(digitalRead(limitSwitchOpen)==HIGH && digitalRead(limitSwitchClosed)==HIGH){ // The curtain is either in Open or closed position at boot time
            IsMoving = 1; // Send command to fully open the curtains
            State->setVal(1);   // 1 means opening
        }
    }

    void loop(){  // loop() method
        if(IsMoving == 1){ // Fully opening the window.
            #ifdef StepperSupport
                stepperMotor->setSpeed(100);
                stepperMotor->step(1);
            #endif
            #ifdef AccelStepperSupport
                if(stepperMotor->distanceToGo() == 0){
                    stepperMotor->setCurrentPosition(0);
                    stepperMotor->moveTo(endPoint);
                }
            #endif
            #ifdef FastAccelStepperSupport
                if(stepperMotor->isRunning() && digitalRead(dirPinStepper)==LOW){
                    stepperMotor->stopMove();
                    while(stepperMotor->isRunning()){}
                }
                stepperMotor->runForward();
            #endif
            LOG1("Opening...\n");
        }

        if(digitalRead(limitSwitchOpen)==LOW && IsMoving == 1){ // Reached the limit of opened position
            #ifdef StepperSupport

            #endif
            #ifdef AccelStepperSupport
                stepperMotor->setCurrentPosition(0);
                stepperMotor->stop();
                stepperMotor->disableOutputs();
            #endif
            #ifdef FastAccelStepperSupport
                stepperMotor->stopMove();
            #endif
            IsMoving = 0;
            Current->setVal(100); // Display curtains as fully opened
            State->setVal(2);   // 2 means stopped
            LOG1("Curtains fully open\n");
        }

        if(IsMoving == 2){ // Fully closing the curtains.
            #ifdef StepperSupport
                stepperMotor->setSpeed(100);
                stepperMotor->step(-1);
            #endif
            #ifdef AccelStepperSupport
                if(stepperMotor->distanceToGo() == 0){
                    stepperMotor->setCurrentPosition(0);
                    stepperMotor->moveTo(-endPoint);
                    Serial.println("endpoint reached");
                }
            #endif
            #ifdef FastAccelStepperSupport
                if(stepperMotor->isRunning() && digitalRead(dirPinStepper)==HIGH){
                    stepperMotor->stopMove();
                    while(stepperMotor->isRunning()){}
                }
                stepperMotor->runBackward();
            #endif
            LOG1("Closing...\n");
        }

        if(digitalRead(limitSwitchClosed)==LOW && IsMoving == 2){ // Reached the limit of closed position
            #ifdef StepperSupport

            #endif
            #ifdef AccelStepperSupport
                stepperMotor->setCurrentPosition(0);
                stepperMotor->stop();
                stepperMotor->disableOutputs();
            #endif
            #ifdef FastAccelStepperSupport
                stepperMotor->stopMove();
            #endif
            IsMoving = 0;
            Current->setVal(0); // Display curtains as fully closed
            State->setVal(2);   // 2 means stopped
            LOG1("Curtains fully closed\n");
        }

        if(digitalRead(limitSwitchClosed)==LOW && digitalRead(limitSwitchOpen)==LOW){ //Treading chain
            #ifdef StepperSupport
                stepperMotor->setSpeed(100);
                stepperMotor->step(1);
            #endif
            #ifdef AccelStepperSupport
                if(stepperMotor->distanceToGo() == 0){
                    stepperMotor->setCurrentPosition(0);
                    stepperMotor->moveTo(endPoint);
                }
            #endif
            #ifdef FastAccelStepperSupport
                if(stepperMotor->isRunning() && digitalRead(dirPinStepper)==LOW){
                    stepperMotor->stopMove();
                    while(stepperMotor->isRunning()){}
                }
                stepperMotor->runForward();
                while(digitalRead(limitSwitchClosed)==LOW && digitalRead(limitSwitchOpen)==LOW){}
                stepperMotor->stopMove();
            #endif
            LOG1("Treading chain...\n");
        }

        #ifdef AccelStepperSupport
            stepperMotor->run();
        #endif
    } // loop
};  // Struct
```
If anyone has managed to make a stepper motor work with Homespun, I would love some help.

Thanks!!!

- Blaise

**Comment by blaisebarrette** [comment]
Hi, 
I could never get a stepper to work with Homespan but good news! I got the AccelStepper library to work flawlessly on a remote Spanpoint accessory!
I’m quite happy to have a esp32 as a hub managing all the remote curtains accessories in the house.
Maybe I’ll also connect a temperature sensor to the hub as a bonus accessory.

for now, the curtains accessories will be connected to the main for supply but I might eventually experiment on a battery operated prototype. I ordered a few 18650 batteries and some charging/step up modules from AliExpress to experiment with.

I found Spanpoint quite easy to work with. Great work again!

Here is a render of the unit I’m working on as a teaser! 🤪

![a3faff1e-2767-4090-8086-55be8179fea6](https://user-images.githubusercontent.com/53702710/200220235-517a675f-dd33-4aa0-8491-43ca026c6082.PNG)

Follow this link to see it in 3D and download if you want: https://a360.co/3Uxdtza

**Reply by HomeSpan** [reply]
That's terrific - thanks for sharing and testing out SpanPoint.

At some point I may create a dedicated class for stepper motors using separate task threads that would work well with HomeSpan, but that's a ways off.
